---
title: "BMS353 Bioinformatics for Biomedical Science - Week 6"
author: "Module Coordinator Mark Dunning"
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
    css: stylesheets/styles.css
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

In the early days of microarray analysis, people were happy if they got a handful of differentially-expressed genes that they could validate or follow-up. However, with later technologies (and depending on the experimental setup) we might have thousands of statistically-significant results, which no-one has the time to follow-up. Also, we might be interested in pathways / mechanisms that are altered and not just individual genes.

In this section we move towards discovering if our results are ***biologically significant***. Are the genes that we have picked statistical flukes, or are there some commonalities. 

There are two different approaches one might use, and we will cover the theory behind both


There is also a bunch of websites for doing the tests
    + we will show how they are done in Bioconductor so the theory is clear

# Incorporating pathway data


We previously annotated our differential expression results to include gene names and description. The same annotation packages will also give us the names of the pathway(s) that each gene belongs to.

```{r}
results_annotated <- readRDS("Robjects/results_cellType_annotated.rds")
```

Let's first filter our data to the genes that have statistical significance.

```{r}
de_genes <- results_annotated %>% 
  filter(padj < 0.05, abs(log2FoldChange) > 1)
```

The organism package allows us to map between different identifiers

```{r}
library(org.Mm.eg.db)
```


Let's remind ourselves of the mappings available as part of this package.

```{r}
columns(org.Mm.eg.db)
```

We can therefore find the pathways that are present amongst this list of genes using the `GO` column

```{r}
de_GO <- AnnotationDbi::select(org.Mm.eg.db, keys = de_genes$ENSEMBL,
                      keytype = "ENSEMBL",
                      columns = "GO")
```

```{r}
dplyr::count(de_GO, GO) 
```

```{r}
dplyr::count(de_GO, GO) %>% 
  arrange(desc(n))
```


## Exercise

<div class="exercise">
- Use the `AnnotationDbi::select` function to find out how many genes exist in total for the GO term `GO:0016020`.
- What proportion of genes for this term were found to be differentially-expressed?
- Is this enough information to decide whether the pathway is altered or not? What information are we not taking into account?
</div>


# Gene Set Testing

```{r}
pathway_genes <- AnnotationDbi::select(org.Mm.eg.db,
                                       keys = "GO:0016020",
                                       keytype = "GO",
                                       columns="SYMBOL") %>% pull(SYMBOL)
```


```{r}
go_table <- mutate(results_annotated, 
                   inPathway = SYMBOL %in% pathway_genes,
                   isDE = padj < 0.05 & abs(log2FoldChange) > 1)
go_table
table(go_table$inPathway, go_table$isDE)
```

```{r}
chisq.test(table(go_table$inPathway, go_table$isDE))
```
    

## Analysis with clusterProfiler

`clusterProfiler` is another Bioconductor package for over-representation analysis. It's main advantage is that it provides some nice visualisation methods.

Firstly, we can identify over-represented GO terms and visualise these as a network.

```{r message=FALSE, warning=FALSE}
library(clusterProfiler)
universe <- results_annotated %>% pull(ENTREZID)
sigGenes <- results_annotated %>% 
  filter(padj < 0.05, !is.na(ENTREZID)) %>% pull(ENTREZID)

enrich_go <- enrichGO(
  gene= sigGenes,
  OrgDb = org.Mm.eg.db,
  keyType = "ENTREZID",
  ont = "BP",
  universe = universe,
  qvalueCutoff = 0.05,
  readable=TRUE
)

```

```{r}
enrich_go_tidy <- enrich_go %>% 
  slot("result") %>% 
  tibble::as.tibble() 
enrich_go_tidy
```

A dot plot can show us the most enriched pathways, and the size of each.

```{r}
dotplot(enrich_go)
```

```{r}
emapplot(enrich_go)
```

We also perform enrichment of KEGG pathways with the `clusterProfiler` package. To run the analysis we can use the same list of genes as before, and need to know the kegg code for our given organism.

```{r}
search_kegg_organism('mouse', by='common_name')
keg_res <- enrichKEGG(gene=sigGenes, organism="mmu")
```

```{r}
head(keg_res,n=10)
```

If we want to view the network for any pathway, we can use the `browseKEGG` function which will take us to the relevant page on the KEGG website.

We will choose an example of a smaller pathway to make things a bit clearer.

```{r}
browseKEGG(keg_res, 'mmu00190')
```

The `pathview` package will produce a version of this plot (`png`) in your working directory.

```{r}
library(pathview)
logFC <- results_annotated$log2FoldChange
names(logFC) <- results_annotated$ENTREZID

pathview(gene.data = logFC, 
         pathway.id = "mmu00190", 
         species = "mmu", 
         limit = list(gene=5, cpd=1))
```    
    

# fgsea analysis

The fgsea package is a free implementation of the Broad's GSEA software and is described in more detail in the package [vignette](http://www.bioconductor.org/packages/release/bioc/vignettes/fgsea/inst/doc/fgsea-tutorial.html) "fast preranked gene set enrichment analysis (GSEA)":

The GSEA analysis is performed by:

- (i) ranking all genes in the data set based on their correlation to the chosen phenotype,
- (ii) identifying the rank positions of all members of the gene set, and 
- (iii) calculating an enrichment score (ES) that represents the difference between the observed rankings and that which would be expected assuming a random rank distribution.

> "After establishing the ES for each gene set across the phenotype, GSEA reiteratively randomizes the sample labels and retests for enrichment across the random classes. By performing repeated class label randomizations, the ES for each gene set across the true classes can be compared to the ES distribution from the random classes. Those gene sets that significantly outperform iterative random class permutations are considered significant." 

The article describing the original software is available [here](http://www.pnas.org/content/102/43/15545.long) and there is also a [commentary on GSEA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1266131/). 

In addition to the GSEA software the Broad also provide a number of very well curated gene sets for testing against your data - the Molecular Signatures Database (MSigDB). Unfortunately, these are collections of human genes. However, these lists have been translated to mouse equivalents by the Walter+Eliza Hall Institutes Bioinformatics service and made avaialble for download. These gene sets use *Entrez ID* as their identifier.

```{r}
library(fgsea)
```


```{r}
results_annotated <- readRDS("Robjects/results_cellType_annotated.rds")
head(results_annotated)
```

An appealing feature of the method is that it does not require us to impose arbitrary cut-offs on the dataset to decide what is differentially-expressed or not. The steps in producing the input required for GSEA are i) retrieving the ranked statistics ii) naming each one according to Entrez ID.

```{r}
library(dplyr)
gseaInput <- filter(results_annotated, !is.na(ENTREZID)) %>% 
  arrange(stat)
ranks <- pull(gseaInput,stat)
names(ranks) <- gseaInput$ENTREZID
```


The Walter+Eliza Hall Institutes Bioinformatics service have made mouse versions of the MSigDB datasets available for download. This should already be available in the `Robjects` folder.

```{r}
download.file("http://bioinf.wehi.edu.au/MSigDB/v7.1/Mm.h.all.v7.1.entrez.rds", destfile = "Robjects/Mm.h.all.v7.1.entrez.rds")
Mm.H <- readRDS("Robjects/Mm.h.all.v7.1.entrez.rds")
pathways <- Mm.H
```

The analysis is one call to the `fgsea` function. We can automatically exclude any pathways with too many or too few genes.

```{r message=FALSE, warning=FALSE}
library(fgsea)
fgseaRes <- fgsea(pathways, ranks, minSize=15, maxSize = 500)
dim(fgseaRes)
#head(fgseaRes)
```

The results table gives the names of each pathway that was tested and the stats from doing the test. We can make this into a "tidy" object with the following code.

```{r}
fgseaResTidy <- fgseaRes %>%
  as_tibble() %>%
  arrange(desc(NES))

# Show in a nice table:
fgseaResTidy 
```

```{r}
library(ggplot2)
ggplot(fgseaResTidy, aes(reorder(pathway, NES), NES)) +
  geom_col(aes(fill=padj<0.05)) +
  coord_flip() +
  labs(x="Pathway", y="Normalized Enrichment Score",
       title="Hallmark pathways NES from GSEA")
```



The enrichment plot will show where the genes belonging to a particular gene set are towards the top or the bottom of the genelist, and how the *enrichment score* is calculated across the dataset.

Here we show the enrichment plot for the pathway with the most positive enrichment score.

```{r}
plotEnrichment(pathways[["HALLMARK_OXIDATIVE_PHOSPHORYLATION"]],
               ranks)
```

> ## Challenge 1{.challenge}
> 1. What pathway has the most extreme negative enrichment score? How can you identify this pathway from the results table?
> 2. Make the enrichment plot for this pathway

### Visualising pathways using a heatmap

The names of genes involved in particular pathways can be extracted from the `pathways` object. We can then use these genes to make a heatmap, or other visualisation, to show how these genes partition the dataset.

```{r}
### Load the dds object if not present
dds <- readRDS("Robjects/dds.rds")
library(pheatmap)
my_genes <- filter(results_annotated, ENTREZID %in% pathways[["HALLMARK_MYC_TARGETS_V2"]]) %>% 
  pull(ENSEMBL)

vsd <- vst(dds)
mat <- assay(vsd)[my_genes,]

dim(mat)
rownames(sampleinfo) <- sampleinfo$run

pheatmap(mat,
         annotation_col = sampleinfo[,c("Status","CellType")],
         scale="row")
```



# Creating Gene lists to use with an online tool

There are also many online tools that one could use to perform a gene set or ontology analysis. 

- [DAVID](https://david.ncifcrf.gov/)
- [GeneTrail](https://genetrail2.bioinf.uni-sb.de/)
- [GORilla](http://cbl-gorilla.cs.technion.ac.il/)

The tools generally require your input genes lists to be uploaded as a simple text file. In this final challenge, we will create some files that you might use in one of these tools.

### A file containing names of background genes

This file has one column which lists **all the gene names** present in the analysis. Gene Symbols are commonly used, although a tool may accept Ensembl or Refseq names

### A file containing names of significant genes

This file has one column which lists the genes that passed the threshold for statistical significance (e.g. p-value less than 0.05) in your analysis. Gene Symbols are commonly used, although a tool may accept Ensembl or Refseq names




**Original Authors: Belinda Phipson, Anna Trigos, Matt Ritchie, Maria Doyle, Harriet Dashnow, Charity Law**, **Stephane Ballereau, Oscar Rueda, Ashley Sawle**
Based on the course [RNAseq analysis in R](http://combine-australia.github.io/2016-05-11-RNAseq/) delivered on May 11/12th 2016 and modified by Cancer Research Uk Cambridge Centre for the [Functional Genomics Autumn School 2017](https://bioinformatics-core-shared-training.github.io/cruk-autumn-school-2017/)

## Resources and data files

This material has been created using the following resources:  
http://www.statsci.org/smyth/pubs/QLedgeRPreprint.pdf 
http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/99-RNAseq_DE_analysis_with_R.html  
http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
https://bioconductor.github.io/BiocWorkshops/rna-seq-data-analysis-with-deseq2.html

This section also uses code from Stephen Turner's guide to fgsea https://stephenturner.github.io/deseq-to-fgsea/
