---
title: "BMS353 Bioinformatics for Biomedical Science - Week 2"
author: "Module Coordinator Mark Dunning"
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
    css: stylesheets/styles.css
editor_options: 
  chunk_output_type: inline
---
```{r setup, include=FALSE, message=FALSE,warning=FALSE}
knitr::opts_chunk$set(message=FALSE,warning = FALSE)
```
<img src="images/logo-sm.png" style="position:absolute;top:40px;right:10px;" width="200" />


# Introduction

In this tutorial we will get our first look at how to analyse high-throughput biological data in R. The majority of the course will discuss RNA-seq data for gene expression, but in this session we will consider data from an older technology; *microarrays*.

Although considered outdated by many, the prevalence of microarrays in public repositories make them a valuable resource.

You will need to have installed the following packages before starting:-

```{r eval=FALSE}
## You should already have these packages from the exercise in week 1
install.packages("BiocManager")
BiocManager::install("GEOquery")
BiocManager::install("limma")
BiocManager::install("genefilter")

```


## Learning outcomes

- Defining the keys steps in the analysis of high-throughput biological data
    + quality assessment
    + normalisation
    + differential expression
- Outline the fundamentals of statistical testing for biological data
- Demonstrate how familiar R functions can be applied to biological data

# Importing the data

The data from this experiment comprises nine paired tumor/normal colon tissues on Illumina HT12\_v3 gene expression Beadchips. Journals are becoming increasingly strict about making sure that data accompanying a publication are available through a suitable repository. The Gene Expression Omnibus (GEO) is one such resource for biological data.


```{r echo=FALSE,message=FALSE}
library(GEOquery)
library(limma)
```

<div class="information">
These data were generated using Illumina microarrays, but the procedure should be highly similar for other manufacturers. e.g. Affymetrix.
</div>

The [landing page](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE33126) has a link to download the *series matrix*. We can download this within R if we wish using the function `download.file`.

The `getGEO` function from `GEOquery` can then import these data into R. The data are a bit more complicated that the spreadsheets we have dealt with so far, so we have to use a specialised package rather than `read_csv` etc.

```{r message=FALSE}
library(GEOquery)
## can change the url to be the dataset that you want.

url <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE33nnn/GSE33126/matrix/GSE33126_series_matrix.txt.gz"

download.file(url, destfile = "GSE33126_series_matrix.txt.gz")

gse <- getGEO(filename = "GSE33126_series_matrix.txt.gz")

```

## Data Representation

We usually print the contents of an object to the screen by running a line of code corresponding to the name of the object. However, the `gse` object works a bit differently.

```{r}
gse

```

The microarray technology employed has given us a gene expression measurement for each probe (or gene) and biological sample. This is a data matrix with one row for each probe (or gene ) and one column for each sample. However, this is not sufficient information for the analysis. We also need to know

- Which biological groups each sample belongs to, and any other relevant information.
    + e.g. tumour/normal status, age, gender, batch
- Which gene each probe is measuring
    + what is the common name for the gene
    + what are it's genomic coordinates
    + what pathways does it belong to.
    
These different sources of data are all available in the `gse` object and are accessed using different functions.

```{r eval=FALSE}
pData(gse) ## print the sample information
fData(gse) ## print the gene annotation
exprs(gse) ## print the expression data
```


# Normalisation (if required)

Prior to statistical analysis, we will inspect the data to discover if all possible sources of technical variation have been accounted for, and that all samples are of sufficient quality to be included.

The `exprs` function can retrieve the expression values as a data frame; with one column per-sample and one row per-gene.

The `summary` function can then be used to print the distributions.

```{r}
## exprs get the expression levels as a data frame and get the distribution
summary(exprs(gse))
```

It is more convenient to visualise the distributions on a boxplot. However, using the `boxplot` function on the data is not very illuminating. 

```{r}
## use the base boxplot function so we don't have to re-arrange the data
boxplot(exprs(gse))
```

The issue is with the scale of the dataset. Each sample has a handful of genes that are highly-expressed, but the majority are lowly-expressed. A log$_2$ transformation is often applied prior to visualisation. This compresses the data, make it easier to compare samples. It also has some other nice properties that are helpful for statistical analysis.

The `boxplot` can also assessed to see if the data have been normalised. Each "box" on the plot shows the distribution of gene expression measurements for a particular sample. If so, the distributions of each sample should be highly similar. Data downloaded from public repositories such as GEO will often be normalised.

```{r}
normData <- log2(exprs(gse))

# the parameter outline=FALSE stops outliers being displayed on the plot - making it a bit tidier
boxplot(normData,outline=FALSE)
```

<div class="information">
An easy to follow demonstration of the most-popular normalisation method (quantile normalisation) can be found here
[https://youtu.be/ecjN6Xpv6SE](https://youtu.be/ecjN6Xpv6SE)
</div>

## Inspect the clinical variables

Data submitted to GEO contain sample labels assigned by the experimenters, and some information about the processing protocol. All these data can be extracted by the `pData` function. 

**For your own data, you will have to decide which columns will be useful in the analysis**. This will include the column giving the main comparison(s) of interest and any potential confounding factors. In this particular dataset it looks like `source_name_ch1` and `characteristics_ch1.1`.

We can use the `select` function from `dplyr` to display just these columns of interest. At this stage it will also be useful to rename the columns to something more convenient using the `rename` function.

```{r}
library(dplyr)
sampleInfo <- pData(gse)
sampleInfo

## source_name_ch1 and characteristics_ch1.1 seem to contain factors we might need for the analysis. Let's pick just those columns

sampleInfo <- select(sampleInfo, source_name_ch1,characteristics_ch1.1)

## Optionally, rename to more convenient column names
sampleInfo <- rename(sampleInfo,group = source_name_ch1, patient=characteristics_ch1.1)
```

Our sample information is therefore:-

```{r}
sampleInfo
```


## Incorporating the feature data

We can export the expression data to a `csv` for inspection in Excel using the `write_csv` function from `readr`. The expression values themselves will probably not be very useful as they will be named according to manufacturer ID rather than gene name (for example). We can create a matrix by joining the expression matrix with the feature annotation.

```{r}
library(readr)
# cbind:- column bind. Join two matrices together

full_output <- cbind(fData(gse),normData)
write_csv(full_output, path="gse_full_output.csv")
```

The annotation from GEO might contain lots of columns that we are not particularly interested in. To keep the data tidier we can use the `select` function to only print particular columns in the output.

```{r}
features <- fData(gse)
View(features)
### Look at the features data frame and decide the names of the columns you want to keep
features <- select(features,Symbol,Entrez_Gene_ID,Chromosome,Cytoband)
full_output <- cbind(features,exprs(gse))
write_csv(full_output, path="gse_full_output.csv")

```


# Statistical Testing

By far the most-popular package for performing differential expression is `limma`. The user-guide is extensive and covers the theory behind the analysis and many use-cases (Chapters 9 and 17 for single-channel data such as Illumina and Affymetrix)

https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf

To illustrate the process of statistical testing, we will use a simple t-test, **which is not recommended in practice**.

To motivate the problem we are trying to solve, we can look at the distribution of expression values for a single gene using a boxplot. We use the base R convention for subsettting a data frame; which is to use square brackets (`[]`) with a row and column index to indicate the rows and columns of data we want to retrieve. If the column index is omitted, it means show all columns. 

The `~` notation is used to specify the sample groups we want to compare between. Although `ggplot` is recommended for publication quality graphics, it is sometimes easier to use the *base* equivalent.

```{r}
fac <- as.factor(sampleInfo$group)
boxplot(normData[1,]~fac)
```
It would appear that the expression level is higher in normal samples compared to tumor. We can perform a statistical test to assess the difference. If the data can be assumed to be *normally-distributed* and *independent* then a two-sample t-test can be employed yielding a p-value and measure of the difference between the sample groups.

```{r}
t.test(normData[1,]~fac,var.equal=TRUE)
```

It would be tedious to test each gene individually. Fortunately the `genefilter` package includes a convenient `rowttests` function that will rapidly test each gene. The output shows the test statistic, difference of means (dm) and p-value for each gene.


```{r}
library(genefilter)
fac <- as.factor(sampleInfo$group)
test_results <- rowttests(normData, fac)
head(test_results)
```
However, you might notice from the output that the names of each probe doesn't have a column heading associated with it. We have to make a small change to the output so that `dplyr` functions can be used.

```{r}
## dplyr doesn't like rownames. Instead add the probe IDs to a column in the data frame

test_results <- tibble::rownames_to_column(test_results,"ID")
```

## Multiple testing correction

Unfortunately, generating the p-values is not the end of the story. Since genome-wide technologies permit us to interrogate many thousands of genes simultaneously, we are likely to obtain many *false-positive* findings. The concept of false positives is nicely summarised in this graphic:-

![](https://raw.githubusercontent.com/sheffield-bioinformatics-core/IntroductionToStats/master/images/EffectSizeFAQs.PNG)

[original source for picture](https://effectsizefaq.com/2010/05/31/i-always-get-confused-about-type-i-and-ii-errors-can-you-show-me-something-to-help-me-remember-the-difference/)

Rather than reporting the *raw* p-values it is highly-recommended to adjust for the fact that we performed many thousands of tests. The most popular method is the Benjamini-Hochberg (BH) as described in the following video. 


<div class="information">

StatQuest video on multiple testing

[https://youtu.be/K8LQSvtjcEo](https://youtu.be/K8LQSvtjcEo)
</div>

```{r}
sum(test_results$p.value < 0.05)
nrow(test_results)*0.05 
```

Now, assume for a moment that the null hypothesis is true for all genes, i.e., no gene is affected by the change between tumour and normal. Then, by the definition of p value, we expect up to 5% of the genes to have a p-value below 0.05. This amounts to 2440 genes. If we just considered the list of genes with a p value below 0.05 as differentially expressed, this list should therefore be expected to contain up to 2440/8158 = 30% false positives

The `p.adjust` function is a general-purpose function for adjusting p-values.  The Benjamini-Hochberg method calculates for each gene an adjusted p value which answers the following question: if one called significant all genes with a p value less than or equal to this gene’s p value threshold, what would be the fraction of false positives (the false discovery rate, FDR) among them (in the sense of the calculation outlined above)?

```{r}
?p.adjust
```


### Exercise

<div class="exercise">
- How would you re-arrange the rows of`test_results` according to p-value? Which gene has the lowest p-value?
- Add a new column `padj` to `test_results`containing p-values adjusted with the "BH" method. Consult the R introductory course or `dplyr` cheatsheet to recall how to add columns to a data frame.
    + How many genes have an *adjusted* p-value less than 0.05?
</div>

```{r}

```



# Interpretation of Results

At the moment our results are not particularly easy to navigate as the only information to identify each gene is the identifier that the microarray manufacturer has assigned. Fortunately, the GEO entry contains extensive annotation that we can add. The annotation data can be retrieved with the `fData` function and we restrict to columns we are interested in using `select`.

**For your other datasets, you will have to choose the columns that are of interest to you. You probably won't have the same column headings used here**.

Once an annotation data frame has been created, it merged to our results using the `left_join` function.

```{r}
anno <- fData(gse)
anno <- select(anno,Symbol,Entrez_Gene_ID,Chromosome,Cytoband)
anno <- tibble::rownames_to_column(anno,"ID")
test_results <- left_join(test_results,anno)
```


The `filter` function from `dplyr` gives a convenient way to interrogate the table of results.

```{r}
## Get the results for particular gene of interest
filter(test_results, Symbol == "SMOX")
## Get results for genes with TP53 in the name
filter(test_results, grepl("TP53", Symbol))
```

We can also filter according to p-value (adjusted) and fold-change cut-offs

```{r}
filter(test_results, p.value < 0.05, abs(dm) > 1)
```

These results can be exported with the `write_csv` function.

```{r}
library(readr)
filter(test_results, p.value < 0.05, abs(dm) > 1) %>%
  write_csv(path="filtered_de_results.csv")
```


## Visualisation

A common plot for displaying the results of a differential expression analysis is a *volcano plot*. It is a *scatter plot* that shows statistical significance and the magnitude of difference between conditions. They are used to identify which genes are the most significant and are also changing by the most amount.

The data we need for the plot is contained in our `test_results` data frame. This basic plot displays a point for every gene, but does not take advantage of some of the other columns in the data frame.

The p-values need to be transformed so that higher on the y-axis indicates a more significant result. A log10 transformation is typically used.

```{r}
## a few of quick examples of how log10 works
log10(100)
log10(1000)
log10(0.1)
log10(0.001)
log10(0.05)

```


```{r}
library(ggplot2)
ggplot(test_results, aes(x = dm , y = -log10(p.value))) + geom_point()
```


An MA-plot is an alternative to the volcano plot with the x-axis being the average expression level of each gene, and the y-axis is the difference between groups.

![](https://upload.wikimedia.org/wikipedia/commons/6/63/Pre-normalization.MAplot.png)

The `rowMeans` function can be used to calculate the means of each row in a numeric matrix. This will be required to construct an MA plot.

### Exercise
<div class="exercise">
- Use the `rowMeans` function to calculate the average expression level for each gene
- Use `mutate` to insert the average expression levels as a new column
- Create the MA-plot using ggplot2

</div>

```{r}



```


# Exercise for next time

<div class="exercise">

- Write a markdown file to document the analysis of the GEO dataset `GSE33126`
- Include the key stages of the analysis in your report
  + downloading the data
  + quality assessment
  + log$_2$ transformation
  + statistical testing
  + export the results
- Use the `arrange` function to re-order the rows of the `test_results` data frame according to chromosome name
- Modify the volcano plot above so that points are coloured according to whether the gene is differentially-expressed or not with adjusted p-value < 0.05
  + HINT: you can use the `mutate` function to add an extra column to `test_results` indicating whether a p-value is less 
</div>

# Disclaimer

<div class="warning">

Remember that the t-test is not recommended in practice for the analysis of microarray data. For alternative methods, see the limma user guide or watch this video

[Microarray Analysis with limma](https://youtu.be/ZRet1oeGiUU)

</div>


