---
title: "BMS353 Week 4"
author: "Mark Dunning"
date: "November 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(ggthemr)
uos_pal <- function(){
  uos_pal <- c("Process_Cyan"=rgb(0,159,218,maxColorValue = 255),
               "Pantone_274"=rgb(31,20,93,maxColorValue = 255),
               "Process_Yellow"=rgb(249,227,0,maxColorValue = 255),
               
               "Pantone_347"=rgb(0,155,72,maxColorValue = 255),
               "Pantone_382"=rgb(190,214,0,maxColorValue = 255),
               "Process_Magenta"=rgb(209,0,116,maxColorValue = 255),
               "Pantone_Orange_021"=rgb(255,88,0,maxColorValue = 255),
               
               "Pantone_512"=rgb(119,33,111,maxColorValue = 255),
               "Pantone_485"=rgb(213,43,30,maxColorValue = 255),
               "Pantone_Black"=rgb(30,30,30,maxColorValue = 255),
               "Pantone_161"=rgb(98,60,27,maxColorValue = 255),
               
               "Pantone_7501"=rgb(219,206,172,maxColorValue = 255),
               "Pantone_343"=rgb(3,86,66,maxColorValue = 255),
               "Pantone_322"=rgb(0,116,122,maxColorValue = 255),
               "Pantone_202"=rgb(130,36,51,maxColorValue = 255)
  )
  
  uos_pal
}

uos_colours <- as.character(uos_pal())
# you have to add a colour at the start of your palette for outlining boxes, we'll use a grey:
uos_colours <- c("#555555", uos_colours)
# remove previous effects:

ggthemr_reset()
# Define colours for your figures with define_palette
uos <- define_palette(
  swatch = uos_colours, # colours for plotting points and bars
  gradient = c(lower = uos_colours[1L], upper = uos_colours[2L]), #upper and lower colours for continuous colours
  background = "white" #defining a grey-ish background 
)
# set the theme for your figures:
ggthemr(uos)
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(DESeq2))
```

## Week 3 recap

- first glimpse of our RNA-seq dataset
- used QA tools to un-cover sample swaps and annotation issues


## Week 4 overview

## Data Transformation

- previously we transformed the data so we could visualise
- some 

## Data Transformation and normalisation

![](seq_depth.PNG)

(image from Harvard Chan Bioinformatics Core - https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)

## Data Transformation and normalisation

![](gene_length.PNG)

(image from Harvard Chan Bioinformatics Core)

## Data Transformation and normalisation

![](lib_size.PNG)

(image from Harvard Chan Bioinformatics Core - https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html)

## The DESeq pipeline

- DESeq2 provides a one-step method to carry out the required normalised and detect differential expression
  + `DESeq`
  + we can run the steps manually if we wish..
- the comparisons performed are determined by the `design`

```{r echo=FALSE,message=FALSE,warning=FALSE}
de <- DESeq(dds)
de
```


## Design specification

- R's `~` notation is used to define the *design* of the experiment
- for simple cases `~condition` means compare conditions in our dataset
  + `condition` has to be a column in our *meta data*

```{r echo=FALSE,message=FALSE,warning=FALSE}
design(dds)
colData(dds)
```


## Printing results

- results are in a familiar format
- adjusted p-values are automatically calculated
- can 

```{r}
results(de)[1:5,]
```

